File: 1122Programming/leetcode/206. Reverse Linked List/ans.rs
// https://leetcode.com/problems/reverse-linked-list/description
// Problem ID: 206. Reverse Linked List
//
// Author: revival0728
//
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut ptr: Option<Box<ListNode>> = head;
        let mut rev: Option<Box<ListNode>> = None;

        while let Some(cur) = ptr {
            let mut parent = ListNode::new(cur.val);
            parent.next = rev.take();
            rev = Some(Box::new(parent));
            ptr = cur.next;
        }
        
        rev
    }
}

File: 1122Programming/leetcode/206. Reverse Linked List/ans.cpp
/**
 * https://leetcode.com/problems/reverse-linked-list/description
 * Problem ID: 206. Reverse Linked List
 *
 * Author: Ateto
 * 
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = NULL;
        while(head != NULL) {
            ListNode* nxt = head->next;
            head->next = prev;
            prev = head;
            head = nxt;
        }
        return prev;
    }
};